File: lab-3/task-1.1/parrot.c
#include "main.h"

#include <stdlib.h>

char const* name(void* this);
char const* greet();
char const* menu();

typedef struct {
	PTRFUN* vtable;
	char const* name;
} Parrot;

PTRFUN parrot_vtable[] = {name, greet, menu};

int size() {
	return sizeof(Parrot);
}

void construct(void* mem, char const* name)
{
	Parrot* parrot = (Parrot*) mem;

	parrot->vtable = parrot_vtable;
	parrot->name = name;
}

void* create(char const* name)
{
	Parrot* parrot = (Parrot*) malloc(sizeof(Parrot));
	construct(parrot, name);

	return parrot;
}

char const* name(void* this)
{
	Parrot* parrot = (Parrot*) this;
	return parrot->name;
}

char const* greet()
{
	return "gawk";
}

char const* menu()
{
	return "seeds";
}

File: lab-3/task-1.4/src/main/java/hr/fer/zemris/ooup/lab3/model/plugins/Parrot.java
package hr.fer.zemris.ooup.lab3.model.plugins;

import hr.fer.zemris.ooup.lab3.model.Animal;

public class Parrot extends Animal {

    private final String animalName;

    public Parrot(String animalName) {
        this.animalName = animalName;
    }

    @Override
    public String name() {
        return animalName;
    }

    @Override
    public String greet() {
        return "gawk";
    }

    @Override
    public String menu() {
        return "seeds";
    }
}
